### PR TITLE
:memo: livecoding / references and identifiers

### PR DESCRIPTION
**Pull request by fabienjuif**
Preview : https://github.com/Dedetat/pimp-my-powerpuff/tree/references-identifiers#references--identifiers

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -138,8 +138,20 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
 
 ### References & Identifiers
 #### [#39](https://github.com/Dedetat/pimp-my-powerpuff/pull/39/files) - Creating a reference
+> We have a centralized Store which contains a list of Powerpuffs, we can edit one from its index in the array, what if the edited one is always store.edited ?
+ 1. We create a new field named `edited` which contains the reference to the Powerpuff that is edited 💡. We can think of it like the reference an editing screen can work with
+ 2. This field is typed `types.reference` of the `Powerpuff` model
+ 3. We create an action to set this field, go on and call it `setEdited`!
+ 4. Set the first powerpuff as the one that is edited 👍
```
Comment (delphinemillet) [inline README.md]: as the edited one

File: README.md [hunk 2]
```diff
@@ -138,8 +138,20 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
 
 ### References & Identifiers
 #### [#39](https://github.com/Dedetat/pimp-my-powerpuff/pull/39/files) - Creating a reference
+> We have a centralized Store which contains a list of Powerpuffs, we can edit one from its index in the array, what if the edited one is always store.edited ?
+ 1. We create a new field named `edited` which contains the reference to the Powerpuff that is edited 💡. We can think of it like the reference an editing screen can work with
```
Comment (delphinemillet) [inline README.md]: You can see it like the reference which an editing screen can work with

File: README.md [hunk 3]
```diff
@@ -138,8 +138,20 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
...
+ 2. This field is typed `types.reference` of the `Powerpuff` model
+ 3. We create an action to set this field, go on and call it `setEdited`!
+ 4. Set the first powerpuff as the one that is edited 👍
+ 5. Print the Store snapshot
+ 6. Oups, it doesn't work 🙀
 
 #### [#40](https://github.com/Dedetat/pimp-my-powerpuff/pull/40/files) - Reference an identifier
+> A reference need an identifier
```
Comment (delphinemillet) [inline README.md]: needs

File: README.md [hunk 4]
```diff
@@ -138,8 +138,20 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
...
+  1. We should add an `identifier` to our Powerpuff, a number is fine, let's call it... hm... `id` 😎 !!
```
Comment (delphinemillet) [inline README.md]: we must

File: README.md [hunk 5]
```diff
@@ -138,8 +138,20 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
 
 ### References & Identifiers
 #### [#39](https://github.com/Dedetat/pimp-my-powerpuff/pull/39/files) - Creating a reference
+> We have a centralized Store which contains a list of Powerpuffs, we can edit one from its index in the array, what if the edited one is always store.edited ?
+ 1. We create a new field named `edited` which contains the reference to the Powerpuff that is edited 💡. We can think of it like the reference an editing screen can work with
+ 2. This field is typed `types.reference` of the `Powerpuff` model
+ 3. We create an action to set this field, go on and call it `setEdited`!
+ 4. Set the first powerpuff as the one that is edited 👍
+ 5. Print the Store snapshot
```
Comment (delphinemillet) [inline README.md]: `Store` snapshot

File: README.md [hunk 6]
```diff
@@ -138,8 +138,20 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
...
+ 6. Oups, it doesn't work 🙀
 
 #### [#40](https://github.com/Dedetat/pimp-my-powerpuff/pull/40/files) - Reference an identifier
+> A reference need an identifier
+  1. We should add an `identifier` to our Powerpuff, a number is fine, let's call it... hm... `id` 😎 !!
+  2. The `id` is now required, we add it to `preProcessSnapshot` as a random number (🤢, this is just for the demo) and to the `belle` Powerpuff instance as an arbitary number
+  3. Let's try to print the `Store` snapshot now!
+  4. 🎉 it works 🎉, you can see the snapshot doesn't copy the value to the reference (`edited` field), mobx-state-tree take care of optimisations, what a good boy 🐶!
```
Comment (delphinemillet) [inline README.md]: copy the value on the reference (? not sure about it)

Comment (delphinemillet) [reply README.md]: mobx-state-tree takes care

File: README.md [hunk 7]
```diff
@@ -138,8 +138,20 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
 
 ### References & Identifiers
 #### [#39](https://github.com/Dedetat/pimp-my-powerpuff/pull/39/files) - Creating a reference
+> We have a centralized Store which contains a list of Powerpuffs, we can edit one from its index in the array, what if the edited one is always store.edited ?
```
Comment (delphinemillet) [inline README.md]: [...] of Powerpuffs `and` we can edit one from [...]`.` What if [...]

File: README.md [hunk 8]
```diff
@@ -138,8 +138,20 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
 
 ### References & Identifiers
```
Comment (delphinemillet) [inline README.md:139]: add a short sentence to introduce the topic

**Review by delphinemillet — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (fabienjuif) [reply README.md]: `doesn't copy the value into `edited` field (which is a reference)`

File: README.md [hunk 9]
```diff
@@ -137,9 +137,23 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
 
 
 ### References & Identifiers
+> We have a centralized Store which contains a list of Powerpuffs and we can edit one from its index in the array. What if the edited one is always store.edited ?
+
 #### [#39](https://github.com/Dedetat/pimp-my-powerpuff/pull/39/files) - Creating a reference
+> First of all we add the `edited` field and declare it as a reference
+ 1. We create a new field named `edited` which contains the reference to the Powerpuff that is edited 💡. You can see it like the reference which an editing screen can work with
```
Comment (delphinemillet) [inline README.md]: to the currently edited Powerpuff

Comment (delphinemillet) [reply README.md]: You can see it like `a` reference
(sorry about that)

File: README.md [hunk 10]
```diff
@@ -137,9 +137,23 @@ Welcome to our source speech repository about [mobx-state-tree](https://github.c
...
+ 4. Set the first powerpuff as the edited one 👍
+ 5. Print the `store` snapshot
+ 6. Oups, it doesn't work 🙀
 
 #### [#40](https://github.com/Dedetat/pimp-my-powerpuff/pull/40/files) - Reference an identifier
+> A reference needs an identifier
+  1. We must add an `identifier` to our Powerpuff, a number is fine, let's call it... hm... `id` 😎 !!
+  2. The `id` is now required, we add it to `preProcessSnapshot` as a random number (🤢, this is just for the demo) and to the `belle` Powerpuff instance as an arbitary number
+  3. Let's try to print the `Store` snapshot now!
+  4. 🎉 it works 🎉, you can see the snapshot doesn't copy the value into `edited` field (which is a reference), mobx-state-tree takes care of optimisations, what a good boy 🐶!
```
Comment (delphinemillet) [inline README.md]: `**mobx-state-tree**`

**Review by delphinemillet — CHANGES_REQUESTED** [review]

**Review by delphinemillet — APPROVED** [review]

**Pull request merged**